src/Compiler/Utilities/Activity.fs

```diff
@@ -23,3 +55,81 @@ module Activity =
         activity
 
     let startNoTags (name: string) : IDisposable = activitySource.StartActivity(name)
+
+    let private escapeStringForCsv (o: obj) =
+        if isNull o then
+            ""
+        else
+            let mutable txtVal = o.ToString()
+            let hasComma = txtVal.IndexOf(',') > -1
+            let hasQuote = txtVal.IndexOf('"') > -1
+
+            if hasQuote then
+                txtVal <- txtVal.Replace("\"", "\\\"")
+
+            if hasQuote || hasComma then
+                "\"" + txtVal + "\""
+            else
+                txtVal
+
+    let private createCsvRow (a: Activity) =
+        let sb = new StringBuilder(128)
+
+        let appendWithLeadingComma (s: string) =
+            sb.Append(',') |> ignore
+            sb.Append(s) |> ignore
+
+        // "Name,StartTime,EndTime,Duration,Id,ParentId"
+        sb.Append(a.DisplayName) |> ignore
+        appendWithLeadingComma (a.StartTimeUtc.ToString("HH-mm-ss.ffff"))
+        appendWithLeadingComma ((a.StartTimeUtc + a.Duration).ToString("HH-mm-ss.ffff"))
+        appendWithLeadingComma (a.Duration.TotalSeconds.ToString("000.0000"))
+        appendWithLeadingComma (a.Id)
+        appendWithLeadingComma (a.ParentId)
+
+        let rec rootID (act: Activity) =
```

Comment:
Isn't Acitivity.RootId what we already need?
https://source.dot.net/#System.Diagnostics.DiagnosticSource/System/Diagnostics/Activity.cs,284

Reply:
I don't like what it does with string-based navigation https://source.dot.net/#System.Diagnostics.DiagnosticSource/System/Diagnostics/Activity.cs,775d64ba2a5d0d39 , feels strange and allocate-y, compared to following one or few reference pointers.

Reply:
I leave it to you. It looks like the impl above deals with some edge cases which we might or might not want to take into consideration